services/director-v2/src/simcore_service_director_v2/modules/comp_scheduler/base_scheduler.py

```diff
@@ -0,0 +1,339 @@
+                    project_id,
+                    iteration,
+                    pipeline_params.mark_for_cancellation,
+                )
+                for (
+                    user_id,
+                    project_id,
+                    iteration,
+                ), pipeline_params in self.scheduled_pipelines.items()
+            ]
+        )
+
+    async def _get_pipeline_dag(self, project_id: ProjectID) -> nx.DiGraph:
+        comp_pipeline_repo: CompPipelinesRepository = get_repository(
+            self.db_engine, CompPipelinesRepository
+        )  # type: ignore
+        pipeline_at_db: CompPipelineAtDB = await comp_pipeline_repo.get_pipeline(
+            project_id
+        )
+        pipeline_dag = pipeline_at_db.get_graph()
+        return pipeline_dag
+
+    async def _get_pipeline_tasks(
+        self, project_id: ProjectID, pipeline_dag: nx.DiGraph
+    ) -> Dict[str, CompTaskAtDB]:
+        comp_tasks_repo: CompTasksRepository = get_repository(
+            self.db_engine, CompTasksRepository
+        )  # type: ignore
+        pipeline_comp_tasks: Dict[str, CompTaskAtDB] = {
+            str(t.node_id): t
+            for t in await comp_tasks_repo.get_comp_tasks(project_id)
+            if (str(t.node_id) in list(pipeline_dag.nodes()))
+        }
+        if len(pipeline_comp_tasks) != len(pipeline_dag.nodes()):
+            raise InvalidPipelineError(
+                f"{project_id}"
+                f"The tasks defined for {project_id} do not contain all the tasks defined in the pipeline [{list(pipeline_dag.nodes)}]! Please check."
+            )
+        return pipeline_comp_tasks
+
+    async def _update_run_result_from_tasks(
+        self,
+        user_id: UserID,
+        project_id: ProjectID,
+        iteration: PositiveInt,
+        pipeline_tasks: Dict[str, CompTaskAtDB],
+    ) -> RunningState:
+
+        pipeline_state_from_tasks: RunningState = get_pipeline_state_from_task_states(
+            list(pipeline_tasks.values()),
+        )
+        await self._set_run_result(
+            user_id, project_id, iteration, pipeline_state_from_tasks
+        )
+        return pipeline_state_from_tasks
+
+    async def _set_run_result(
+        self,
+        user_id: UserID,
+        project_id: ProjectID,
+        iteration: PositiveInt,
+        run_result: RunningState,
+    ) -> None:
+        comp_runs_repo: CompRunsRepository = get_repository(
+            self.db_engine, CompRunsRepository
+        )  # type: ignore
+        await comp_runs_repo.set_run_result(
+            user_id=user_id,
+            project_id=project_id,
+            iteration=iteration,
+            result_state=run_result,
+            final_state=(run_result in COMPLETED_STATES),
+        )
+
+    @abstractmethod
+    async def _start_tasks(
+        self,
+        user_id: UserID,
+        project_id: ProjectID,
+        scheduled_tasks: List[TaskIn],
+        callback: Callable[[], None],
+    ) -> None:
+        pass
```

Comment:
I think all abstractmethods should raise [NotImplementedError](https://docs.python.org/3/library/exceptions.html#NotImplementedError)  (see definition in link), right?

Reply:
as discussed. python already complains with a TypeError and says why. so not sure it is necesary